### DISCUSSION TITLE
MP camera for esp32-s3

### DISCUSSION THREAD

**Discussion by shariltumin** [micropython/micropython#12032] [Show and tell]
Those interested in the esp32-s3 firmware with camera module can have a look [here](https://github.com/shariltumin/esp32-cam-micropython-2022/tree/main/firmwares-20230717/ESP32-S3).

Currently three ESP32-S3 camera boards are supported.

    LILYGO-T-Camera-OV2640 N16R8 (Flash 16MB, PSRAM 8MB)
    XIAO-Sense-OV2640 N8R8 (Flash 8MB, PSRAM 8MB)
    Freenove-OV2640 N8R8 (Flash 8MB, PSRAM 8MB)

**Comment by scruss** [comment]
Thanks for those. Out of curiosity, I tried the LILYGO-T-Camera-OV2640 firmware with a Lilygo [T-SIMCAM](https://www.lilygo.cc/en-ca/products/t-simcam). While it did produce a REPL, unsurprisingly the camera didn't work

**Reply by shariltumin** [reply]
```py
# Lilygo T-SIMCAM Board (12Mhz ???)
t_simcam =    {PIN_PWDN:-1, 
              PIN_RESET:18,
              PIN_XCLK:14, 
              PIN_SIOD:4,
              PIN_SIOC:5,
              PIN_D7:15,
              PIN_D6:16,
              PIN_D5:17,
              PIN_D4:12,
              PIN_D3:10, 
              PIN_D2:8, 
              PIN_D1:9, 
              PIN_D0:11,  
              PIN_VSYNC:6,
              PIN_HREF:7,
              PIN_PCLK:13,
              XCLK_MHZ:12, # not sure what speed
              PIXFORMAT:5,
              FRAMESIZE:10,
              JPEG_QUALITY:10,
              FB_COUNT:1,
} 
```

Please try adding the above settings to the "cam_config.py" file. These values are from the link you provided.

The "cam_config.py" file can be found at https://github.com/shariltumin/esp32-cam-micropython-2022/blob/main/Christmas-2023-Edition/scripts/cam_config.py

Then try this init script:

```py
import camera
import cam_config as cc

# set camera configuration
cc.configure(camera, cc.t_simcam)
camera.init()
```

Hopefully this will work.

**Reply by scruss** [reply]
that did it, thanks!

**Comment by rena2019** [comment]
I bought a cheap ESP32-S3 board with CAM interface, SD card slot (labeled "127294G_Y1127_231016" on the back side) from AliExpress
![S759d24d72eef432db8c8bb783d11f8d2d](https://github.com/micropython/micropython/assets/3251463/657faca0-63f0-4310-8b73-aec6bb3a1c21) with OV5640 cam that looks like the Freenove board. First I tried @shariltumin firmware but got no REPL. Then I tried GENERIC ESP32-S3 Micropython firmware ESP32_GENERIC_S3-SPIRAM_OCT-20240222-v1.22.2.bin with working REPL. Any idea why Freenove firmware does not work?

**Reply by shariltumin** [reply]
Sorry to hear that. The firmware only supports OV2640.

```bash
$ tio /dev/ttyACM0
[tio 14:09:47] tio v1.32
[tio 14:09:47] Press ctrl-t q to quit
[tio 14:09:55] Connected

>>>
>>> import machine
>>> machine.reset()

[tio 13:52:14] Disconnected
[tio 13:52:15] Connected

>>>
MPY: soft reboot
MicroPython v1.20.0-206-g33b403dfb-kaki5 on 2024-01-17; ESP32S3-FREENOVE OV2640 (KAKI5) with ESP32-S3
>>> import camera
>>> camera.init()
True
```

You should get the REPL. Which of the two USB ports you tried? To get to the REPL I connect to the USB-OTG port. Press ENTER a few times to get the prompt.

**Reply by rena2019** [reply]
My fault...I forgot to erase flash before writing the firmware file. Now I get the REPL:

```
MicroPython v1.20.0-206-g33b403dfb-kaki5 on 2023-07-10; ESP32S3-FREENOVE OV2640 w/BLE (KAKI5) with ESP32-S3
>>> import camera
>>> camera.init()
False
```

But how it is possible to configure the OV5640 camera? Or do I have to build a custom Micropython ? @shariltumin where can I find the latest source for the cam module?

**Comment by stephanelsmith** [comment]
How cool!  timing is exciting, I've just been working on an spi only camera [ArduMega SPI camera](https://www.arducam.com/product/presale-mega-5mp-color-rolling-shutter-camera-module-with-autofocus-lens-for-any-microcontroller/) as I've been a bit intimidate by the pin outs.  At least this version, just got it taking pictures last night, working with cameras is quite fun! https://github.com/stephanelsmith/ArducamMega-micropython

Your project seems to be a better way to go, I'm a bit concerned about size/weight for my eventual application.  I will see if I can pick up some of the hardware and try it out.  

Thank you for sharing your project.  Looking forward to checking it out!

**Reply by shariltumin** [reply]
You can get ESP32-Cam with downloader module for US$6.00 from AliExpress.

**Comment by cnadler86** [comment]
Hi guys,
for those hitting this post on the search of a MP-Camera driver: I just released one that works on the current Micropython version and has OV2640 and OV5640 support. At the moment have only precompiled FW for generic boards, but more will follow. Here the [link](https://github.com/cnadler86/micropython-camera-API).

**Reply by shariltumin** [reply]
Very nice Christopher. I suggest you create a new thread so that others can properly send thanks for your efforts and eventually requests for help.

I recommend everyone to try Christopher's version.

**Reply by ricksorensen** [reply]
Great.  Adding it to awesome-micropython (https://awesome-micropython.com/#camera) would be nice too.

**Reply by cnadler86** [reply]
Hi. Thanks shariltumin for the hint. That's a good idea. Will open an new thread then. 

I submitted a PR for awesome MP  yesterday.

**Reply by d4c00** [reply]
You're a lifesaver! This is exactly what I was looking for!

**Comment by Shou-a** [comment]
dear i have OV3660-v1 want to use with ESP32-S3-Cam could you help please

**Reply by cnadler86** [reply]
[https://github.com/cnadler86/micropython-camera-API/releases](https://github.com/cnadler86/micropython-camera-API/releases).
Here you will find precompiled binaries for your camera.
